prov/gni/src/gnix_msg.c

```diff
@@ -558,6 +728,315 @@ static int __gnix_rndzv_req(void *arg)
+
+			req->msg.recv_info[i].mem_hndl =
+				req->msg.recv_md[i]->mem_hndl;
+
+			GNIX_DEBUG(FI_LOG_EP_DATA, "auto-reg MR: %p\n",
+				   req->msg.recv_md[i]);
+
+		}
+		req->msg.recv_flags |= FI_LOCAL_MR;
+	}
+
+	/*
+	 * TODO: "coalesce" the user's sendv buffer before sending the control
+	 * message; do "the same" in _gnix_recvv.
+	 */
+	i = 0;
+
+	/* Iterate through the buffers and build the Fma and Rdma requests! */
+	while (i < send_cnt) {
+		get_len = MIN(recv_len, req->msg.send_info[i].send_len);
+
+		GNIX_DEBUG(FI_LOG_EP_DATA, "send_info[%d].send_len = %lu,"
+			   " recv_len = %lu, get_len = %lu\n", i,
+			   req->msg.send_info[i].send_len, recv_len, get_len);
+
+		if (get_len >= rndzv_thresh) { /* Build the rdma txd */
+			ret = _gnix_nic_tx_alloc(nic, &txd);
+
+			if (ret != FI_SUCCESS) {
+				/* We'll try again. */
+				GNIX_INFO(FI_LOG_EP_DATA, "_gnix_nic_tx_alloc()"
+					  " returned %s\n",
+					  fi_strerror(-ret));
+				return -FI_ENOSPC;
+			}
+
+			txd->completer_fn = __gnix_rndzv_iov_req_complete;
+			txd->req = req;
+			txd->gni_desc.type = GNI_POST_RDMA_GET;
+			txd->gni_desc.cq_mode = GNI_CQMODE_GLOBAL_EVENT;
+			txd->gni_desc.dlvr_mode = GNI_DLVMODE_PERFORMANCE;
+			txd->gni_desc.local_mem_hndl =
+				req->msg.recv_info[j].mem_hndl;
+			/* TODO: If the user passes in a incorrect mdh can we
+			 * report a useful error? */
+			txd->gni_desc.remote_mem_hndl = req->msg.send_info[i].mem_hndl;
+			txd->gni_desc.rdma_mode = 0;
+			txd->gni_desc.src_cq_hndl = (use_tx_cq_blk) ?
+				nic->tx_cq_blk : nic->tx_cq;
+
+			/* TODO: handle alignment! */
+			txd->gni_desc.local_addr =
+				(uint64_t) recv_ptr;
+			txd->gni_desc.remote_addr =
+				req->msg.send_info[i].send_addr;
+			txd->gni_desc.length = get_len;
+
+			slist_insert_tail(&txd->slist, &req->iov_txd_slist);
+			txd_cnt++;
+			txd = NULL;
+		} else {		       /* Build the Ct txd */
+			/*
+			 * The txd is ready to be queued if the max size has
+			 * been reached.
+			 */
+			if (ct_size + get_len >= max_ct_size) {
+				*next_ct = txd = NULL;
+				slist_insert_tail(&txd->slist,
+						  &req->iov_txd_slist);
+				txd_cnt++;
+				ct_size = 0;
+			} else {
+				if (!txd) {
+					GNIX_DEBUG(FI_LOG_EP_DATA, "New FMA"
+						   " CT\n");
+					ret = _gnix_nic_tx_alloc(nic, &txd);
+					if (ret != FI_SUCCESS) {
+						/* We'll try again. */
+						GNIX_INFO(FI_LOG_EP_DATA,
+							  "_gnix_nic_tx_alloc()"
+							  " returned %s\n",
+							  fi_strerror(-ret));
+						return -FI_ENOSPC;
+					}
+
+					txd->completer_fn =
+						__gnix_rndzv_iov_req_complete;
+					txd->req = req;
+					txd->gni_desc.type = GNI_POST_FMA_GET;
+					txd->gni_desc.cq_mode =
+						GNI_CQMODE_GLOBAL_EVENT;
+					txd->gni_desc.dlvr_mode =
+						GNI_DLVMODE_PERFORMANCE;
+					txd->gni_desc.local_mem_hndl =
+						req->msg.recv_info[j].
+						mem_hndl;
+
+					/* TODO: If the user passes in a
+					 * incorrect mdh can we report a
+					 * useful error? */
+					txd->gni_desc.remote_mem_hndl =
```

Comment:
Definitely can't do anything here, and again you should get some ugni error later.